scripts/emulator-testing/emulators/emulator.ts

```diff
@@ -51,35 +50,59 @@ export abstract class Emulator {
     return new Promise<void>((resolve, reject) => {
       tmp.dir((err: Error | null, dir: string) => {
         if (err) reject(err);
-
         console.log(`Created temporary directory at [${dir}].`);
         const filepath: string = path.resolve(dir, this.binaryName);
-        const writeStream: fs.WriteStream = fs.createWriteStream(filepath);
-
+        const writer = fs.createWriteStream(filepath);
         console.log(`Downloading emulator from [${this.binaryUrl}] ...`);
-        fetch(this.binaryUrl).then(resp => {
-          resp.body
-            .pipe(writeStream)
-            .on('finish', () => {
-              console.log(`Saved emulator binary file to [${filepath}].`);
-              // Change emulator binary file permission to 'rwxr-xr-x'.
-              // The execute permission is required for it to be able to start
-              // with 'java -jar'.
-              fs.chmod(filepath, 0o755, err => {
-                if (err) reject(err);
-                console.log(
-                  `Changed emulator file permissions to 'rwxr-xr-x'.`
-                );
-                this.binaryPath = filepath;
-
-                if (this.copyToCache()) {
-                  console.log(`Cached emulator at ${this.cacheBinaryPath}`);
+        // Map the DOM's fetch Reader to node's streaming file system
+        // operations. We will need to access class members `binaryPath` and `copyToCache` after the
+        // download completes. It's a compilation error to pass `this` into the named function
+        // `readChunk`, so the download operation is wrapped in a promise that we wait upon.
+        const downloadPromise = new Promise<void>(
+          (downloadComplete, downloadFailed) => {
+            fetch(this.binaryUrl)
+              .then(resp => {
+                if (resp.status !== 200) {
+                  console.log('Download of emulator failed: ', resp.statusText);
+                  downloadFailed();
                 }
-                resolve();
+                const reader = resp.body?.getReader();
+                reader?.read().then(function readChunk({ done, value }): any {
```

Comment:
Is there a chance of this hanging if there's no reader or no body? Do we have to call a resolve() or reject() or downloadFailed() or something at the end as a failsafe?

Reply:
Yes, thank you! Updated the code.